skrifa/src/outline/glyf/hint/definition.rs

```diff
@@ -0,0 +1,230 @@
+        if let Some(def) = defs.get(key as usize) {
+            if def.is_active() && def.key == key {
+                return Ok(def);
+            }
+        }
+        // Otherwise, walk backward doing a linear search.
+        for def in defs.iter().rev() {
+            if def.is_active() && def.key == key {
+                return Ok(def);
+            }
+        }
+        Err(HintErrorKind::InvalidDefintion(key as _))
+    }
+
+    /// Returns a reference to the underlying definition slice.
+    pub fn as_slice(&self) -> &[Definition] {
+        match self {
+            Self::Ref(defs) => defs,
+            Self::Mut(defs) => defs,
+        }
+    }
+
+    /// If the map is mutable, resets all definitions to the default
+    /// value.
+    pub fn reset(&mut self) {
+        if let Self::Mut(defs) = self {
+            defs.fill(Default::default())
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn alloc(set: &mut DefinitionMap, key: i32) {
+        let def = set.allocate(key).unwrap();
+        *def = Definition::new(Program::Font, 0..0, key);
+    }
+
+    #[test]
+    fn too_many_and_invalid() {
+        let mut buf = vec![Default::default(); 32];
+        let mut map = DefinitionMap::Mut(&mut buf);
+        for i in 0..32 {
+            map.allocate(i).unwrap();
+        }
+        assert!(matches!(
+            map.allocate(33),
+            Err(HintErrorKind::TooManyDefinitions)
+        ));
+        assert!(matches!(
+            map.get(33),
+            Err(HintErrorKind::InvalidDefintion(33))
+        ));
+    }
+
+    #[test]
+    fn allocate_dense() {
+        let mut buf = vec![Default::default(); 32];
+        let mut map = DefinitionMap::Mut(&mut buf);
+        for i in 0..32 {
+            map.allocate(i).unwrap();
+        }
+        for (i, def) in map.as_slice().iter().enumerate() {
+            let key = i as i32;
+            map.get(key).unwrap();
+            assert_eq!(def.key(), key);
+        }
+    }
+
+    #[test]
+    fn allocate_sparse() {
+        let mut buf = vec![Default::default(); 10];
+        let mut map = DefinitionMap::Mut(&mut buf);
+        // The first 4 keys are in order which should be allocated to an entry
+        // where index == key. The next three will be allocated from the end
+        // of the definition storage. The last one will also be allocated to
+        // its own index because it happens to be free.
+        let keys = [0, 1, 2, 3, 123456, -42, -5555, 5];
+        for key in keys {
+            map.allocate(key).unwrap();
+        }
+        let slice = map.as_slice();
+        // For first 4 and last, index == key
+        for i in (0..4).chain(Some(5)) {
+            assert_eq!(slice[i].key, i as i32);
+        }
+        // The rest would be allocated from the end of the array in
+        // reverse order.
+        for (&a, b) in keys[4..7].iter().rev().zip(&slice[slice.len() - 3..]) {
+            assert_eq!(a, b.key());
+        }
+    }
+}
```

Comment:
it's hard to tell at a glance if we are testing both forms of matching (key works vs backtrack required), I think I might have expected a test specifically for that

Reply:
This test is meant to capture the mixed case where some keys are used as indices and some aren’t. It explicitly checks the order in the backing slice to make sure allocation works as expected.

It is missing one case where a key would have worked as an index but prior fallback allocation blocks it. I’ll add that case and an additional loop over keys to ensure `get()` works successfully for all.

Reply:
Added comments describing the allocation tests (including a new one for fully sparse allocation). Renamed this one to `allocate_mixed` and simplified the backing store check. Also added the case where direct mapping would work but is blocked by prior fallback. This should cover all code paths.